src/Guide/Api/Utils.hs

```diff
@@ -89,3 +94,43 @@ instance (HasSwagger api, KnownSymbol name, KnownSymbol desc) =>
              symbolVal (Proxy @desc))
             Nothing
      in toSwagger (Proxy @api) & applyTags [tag]
+
+
+-- | Type to log user's edition with header details
+data RequestDetails = RequestDetails
+  { rdIp      :: Text
+  , rdReferer :: Text
+  , rdUA      :: Text
+  } deriving (Show, Generic)
+
+instance (HasServer api context)
+  => HasServer (Maybe RequestDetails :> api) context where
+
+  type ServerT (Maybe RequestDetails :> api) m = Maybe RequestDetails -> ServerT api m
+
+  hoistServerWithContext _ pc nt s = hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+  route Proxy context subserver = route (Proxy :: Proxy api) context $
+      subserver `addHeaderCheck` withRequest headerCheck
+    where
+      headerCheck :: Request -> DelayedIO (Maybe RequestDetails)
+      headerCheck req = unfoldRequestArgument (Proxy :: Proxy '[Optional, Strict]) errReq errSt meRequestDatails
+        where
+          meRequestDatails :: Maybe (Either Text RequestDetails)
+          meRequestDatails = getCompose $ liftA2 RequestDetails (Compose ip) (Compose referer) <*> Compose ua
+
+          ip :: FromHttpApiData a => Maybe (Either Text a)
+          ip = fmap parseHeader $ lookup "X-Forwarded-For" (requestHeaders req)
+          referer :: FromHttpApiData a => Maybe (Either Text a)
+          referer = fmap parseHeader $ lookup hReferer (requestHeaders req)
+          ua :: FromHttpApiData a => Maybe (Either Text a)
+          ua = fmap parseHeader $ lookup hUserAgent (requestHeaders req)
+
+          errReq = delayedFailFatal err400
+            { errBody = "One of Headers: " <> "Referer, X-Forwarded-For or User-Agent" <> " is required" }
+          errSt e = delayedFailFatal err400
+            { errBody = "Error parsing one of headers "
+           <> "Referer, X-Forwarded-For or User-Agent"
+           <> " failed: "
```

Comment:
List of headers can be generalized over error messages. 

ALSO, it is a place for future mistake, when we will add a field to RD, but forget to add it here, so there is another great typecheckable solution, whih is based on previous solution: 

```haskell
{-# LANGUAGE DeriveFunctor, DeriveFoldable, DeriveTraversable #-}

import Data.Functor.Compose
import Data.Foldable
import Data.List (intercalate)

data Error = Error
  deriving Show

type RequestDetails = RequestDetails' String
data RequestDetails' a = RequestDetails
  { ip :: a
  , referer :: a
  , userAgent :: a
  } deriving (Functor, Foldable, Traversable, Show)

-- Just for visibility
type HeaderKey = String

parseHeader :: HeaderKey -> Maybe (Either Error String)
-- parseHeader "ip" = Just (Left Error)
-- parseHeader "ip" = Nothing
parseHeader key = Just (Right key)

headersMap :: RequestDetails' HeaderKey
headersMap = RequestDetails
  { ip = "X-Forwarded-For"
  , referer = "Referer"
  , userAgent = "User-Agent"
  }

requestDetails :: Maybe (Either Error RequestDetails)
requestDetails = getCompose $ traverse (Compose . parseHeader) headersMap
  
headers :: String
headers = intercalate ", " (toList headersMap)

err :: String
err = "One of Headers: " <> headers <> " is required"
```

Reply:
We can't forget to add the field – `liftA3` will stop typechecking.